backend/src/main/kotlin/org/loculus/backend/service/submission/CompressionDictService.kt

```diff
@@ -0,0 +1,128 @@
+package org.loculus.backend.service.submission
+
+import org.jetbrains.exposed.sql.transactions.transaction
+import org.loculus.backend.api.Organism
+import org.loculus.backend.config.BackendConfig
+import org.loculus.backend.service.submission.dbtables.CompressionDictionariesTable
+import org.loculus.backend.service.submission.dbtables.CompressionDictionaryEntity
+import org.loculus.backend.utils.DateProvider
+import org.springframework.stereotype.Service
+import java.security.MessageDigest
+import java.util.concurrent.ConcurrentHashMap
+
+class DictEntry(val id: Int, val dict: ByteArray)
+
+/**
+ * An abstraction over the compression_dictionaries_table.
+ * Caches the contents in memory to avoid repeated DB lookups.
+ */
+@Service
+class CompressionDictService(private val backendConfig: BackendConfig, private val dateProvider: DateProvider) {
+    private val caches: Triple<
+        Map<Pair<String, String>, DictEntry>,
+        Map<String, DictEntry>,
+        ConcurrentHashMap<Int, ByteArray>,
+        > by lazy {
+        // Must be lazy to make sure the caches are only populated after the Flyway migration has run
+        // The Spring bean is created before Flyway runs, i.e. we must not read from the DB when creating this class
+        populateCaches()
+    }
+
+    private val dictCache: Map<Pair<String, String>, DictEntry> get() = caches.first
+    private val unalignedDictCache: Map<String, DictEntry> get() = caches.second
+    private val cacheById: ConcurrentHashMap<Int, ByteArray> get() = caches.third
+
+    private fun populateCaches(): Triple<
+        Map<Pair<String, String>, DictEntry>,
+        Map<String, DictEntry>,
+        ConcurrentHashMap<Int, ByteArray>,
+        > {
+        val dictCache = mutableMapOf<Pair<String, String>, DictEntry>()
```

Comment:
Some logging here would be good - as it runs on startup so would be good to know how long it takes etc